File: haskell/haskell-first-principles/ch04-basic-datatypes/tuple1.hs
fst' :: (a, b) -> a
fst' (a, _) = a

snd' :: (a, b) -> b
snd' (_, b) = b

tupFn :: (Int, [a])
      -> (Int, [a])
      -> (Int, [a])
tupFn (a, b) (c, d) = ((a + c), (b ++ d))

--
-- :load tuple1.hs
-- *Main> tupFn (10, "master") (20, " yoda")
-- (30,"master yoda")
--

File: old-notes/scala/effective-scala/week2/tuple_1.scala
package tuple1

type Name = String
type Skill = String
type Level = Int

// A tuple of name and skill
type Jedi = (Name, Skill)

// A three-tuple of name, skill and level.
type Master = (Name, Skill, Level)

val ahsoka: Jedi = ("Ahsoka Tano", "Lightsaber")
val yoda: Master = ("Yoda", "The Force", 100)

//
// Could also use pattern matching like destructuring in EcmaScript:
//
val (ahsokaName, ahsokaSkill) = ahsoka
val (_, _, yodaLevel) = yoda

def getInfo(jedi: Jedi | Master): String =
  jedi match
    case (name, skill) =>
      s"Name: $name, skill: $skill"
    case (name, skill, level) =>
      s"Name: $name, Skill: $skill, Level: $level"

@main def main(): Unit =
  println(getInfo(ahsoka))
  println(getInfo(yoda))
  println(ahsokaName)
  println(yodaLevel)
